cluster-autoscaler/utils/kubernetes/listers.go

```diff
@@ -144,18 +144,63 @@ type PodLister interface {
 	List() ([]*apiv1.Pod, error)
 }
 
+func isScheduled(pod *apiv1.Pod) bool {
+	if pod == nil {
+		return false
+	}
+	return pod.Spec.NodeName != ""
+}
+func isDeleted(pod *apiv1.Pod) bool {
+	if pod == nil {
+		return false
+	}
+	return pod.GetDeletionTimestamp() != nil
+}
+func isUnschedulable(pod *apiv1.Pod) bool {
+	if pod == nil {
+		return false
+	}
+	if isScheduled(pod) || isDeleted(pod) {
+		return false
+	}
+	_, condition := podv1.GetPodCondition(&pod.Status, apiv1.PodScheduled)
+	if condition == nil || condition.Status != apiv1.ConditionFalse || condition.Reason != apiv1.PodReasonUnschedulable {
+		return false
+	}
+	return true
+}
+
 // ScheduledPods is a helper method that returns all scheduled pods from given pod list.
 func ScheduledPods(allPods []*apiv1.Pod) []*apiv1.Pod {
 	var scheduledPods []*apiv1.Pod
 	for _, pod := range allPods {
-		if pod.Spec.NodeName != "" {
+		if isScheduled(pod) {
 			scheduledPods = append(scheduledPods, pod)
 			continue
 		}
 	}
 	return scheduledPods
 }
 
+// SchedulerUnprocessedPods is a helper method that returns all pods which are not yet processed by the scheduler
+func SchedulerUnprocessedPods(allPods []*apiv1.Pod) []*apiv1.Pod {
+	var unprocessedPods []*apiv1.Pod
+	for _, pod := range allPods {
+		// Make sure it's not scheduled or deleted
+		if isScheduled(pod) || isDeleted(pod) || isUnschedulable(pod) {
+			continue
+		}
+		// Make sure that if it's not scheduled it's either
+		// Not processed (condition is nil)
+		// Or Reason is empty (not schedulerError, terminated, ...etc)
+		_, condition := podv1.GetPodCondition(&pod.Status, apiv1.PodScheduled)
+		if condition == nil || (condition.Status == apiv1.ConditionFalse && condition.Reason == "") {
```

Comment:
Do we need this extra check if the pod is neither scheduled, deleted nor unschedulable? It would be useful to have tests for this, especially if the answer is "yes".

Reply:
For Unschedulable, we need the pod scheduling reason to be `PodReasonUnschedulable=Unschedulable`
However, we can have a scheduler error reason `PodReasonSchedulerError=SchedulerError`
In that case, the pod would be neither scheduled, deleted nor unschedulable (according to our definition), but  we don't want to include it as scheduler unprocessed because it's not.
In general our definition of "scheduler unprocessed" means the scheduler has not tried to schedule this pod yet. Which leaves out any non-empty pod scheduling reason